src/Core/Managed/Shared/Extensibility/MetricAggregatorManager.cs

```diff
@@ -0,0 +1,295 @@
+                this.telemetryClient.Flush();
+            }
+            catch (Exception ex)
+            {
+                CoreEventSource.Log.FailedToFlushMetricAggregators(ex.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Takes a snapshot of aggregators collected by this instance of the manager
+        /// and schedules the next snapshot.
+        /// </summary>
+        private Task SnapshotAndReschedule()
+        {
+            return Task.Factory.StartNew(
+                () =>
+                {
+                    try
+                    {
+                        this.Snapshot();
+                    }
+                    catch (Exception ex)
+                    {
+                        CoreEventSource.Log.FailedToSnapshotMetricAggregators(ex.ToString());
+                    }
+                    finally
+                    {
+                        this.snapshotTask.Delay = GetWaitTime();
+                        this.snapshotTask.Start(this.SnapshotAndReschedule);
+                    }
+                });
+        }
+
+        /// <summary>
+        /// Takes snapshot of all active metric aggregators and turns results into metric telemetry.
+        /// </summary>
+        private void Snapshot()
+        {
+            ConcurrentDictionary<MetricAggregator, SimpleMetricStatisticsAggregator> aggregatorSnapshot =
+                Interlocked.Exchange(ref this.aggregatorDictionary, new ConcurrentDictionary<MetricAggregator, SimpleMetricStatisticsAggregator>());
+
+            // calculate aggregation interval duration interval
+            TimeSpan aggregationIntervalDuation = DateTimeOffset.UtcNow - this.lastSnapshotStartDateTime;
+            this.lastSnapshotStartDateTime = DateTimeOffset.UtcNow;
+
+            // prevent zero duration for interval
+            if (aggregationIntervalDuation.TotalMilliseconds < 1)
+            {
+                aggregationIntervalDuation = TimeSpan.FromMilliseconds(1);
+            }
+
+            // adjust interval duration to exactly snapshot frequency if it is close (within 1%)
+            double difference = Math.Abs(aggregationIntervalDuation.TotalMilliseconds - snapshotFrequency.TotalMilliseconds);
+
+            if (difference <= snapshotFrequency.TotalMilliseconds / 100)
+            {
+                aggregationIntervalDuation = snapshotFrequency;
+            }
+
+            if (aggregatorSnapshot.Count > 0)
+            {
+                foreach (KeyValuePair<MetricAggregator, SimpleMetricStatisticsAggregator> aggregatorWithStats in aggregatorSnapshot)
+                {
+                    AggregatedMetricTelemetry aggergatedMetricTelemetry = CreateAggergatedMetricTelemetry(aggregatorWithStats.Key, aggregatorWithStats.Value);
+
+                    aggergatedMetricTelemetry.Properties.Add(intervalDurationPropertyName, ((long)aggregationIntervalDuation.TotalMilliseconds).ToString(CultureInfo.InvariantCulture));
```

Comment:
The `aggregatedMetricTelemetry` can be `null` here. We should add the `intervalDurationPropertyName` inside the `if` statement below.